modules/pico_6lowpan_ll.c

```diff
@@ -0,0 +1,543 @@
+ * Macros
+ ******************************************************************************/
+
+#ifdef DEBUG_6LOWPAN
+#define ll_dbg dbg
+#else
+#define ll_dbg(...) do {} while(0)
+#endif
+
+/*******************************************************************************
+ * Constants
+ ******************************************************************************/
+
+/* Lifetime check interval */
+#define ONE_MINUTE  ((pico_time)(1000 * 60))
+
+/* Number of extensions */
+#define NUM_LL_EXTENSIONS       (2)
+
+/* Possible actions to perform on a received frame */
+#define FRAME_6LOWPAN_LL_RELEASE    (-1)
+#define FRAME_6LOWPAN_LL_DISCARD    (-2)
+
+/*******************************************************************************
+ * Type definitions
+ ******************************************************************************/
+
+typedef struct extension {
+    int (*estimate)(struct pico_frame *f, union pico_ll_addr *src, union pico_ll_addr *dst);
+    int (*out)(struct pico_frame *f, union pico_ll_addr *src, union pico_ll_addr *dst);
+    int (*in)(struct pico_frame *f, union pico_ll_addr *src, union pico_ll_addr *dst);
+} extension_t;
+
+/*******************************************************************************
+ * Global Variables
+ ******************************************************************************/
+
+static struct pico_queue pico_6lowpan_ll_in = {
+    0
+};
+static struct pico_queue pico_6lowpan_ll_out = {
+    0
+};
+
+/*******************************************************************************
+ *  CTX
+ ******************************************************************************/
+
+#ifdef PICO_6LOWPAN_IPHC_ENABLED
+/* Compares if the IPv6 prefix of two IPv6 addresses match */
+static int compare_prefix(uint8_t *a, uint8_t *b, int len)
+{
+    uint8_t bitmask = (uint8_t)(0xff >> (8 - (len % 8)));
+    size_t bytes = (size_t)len / 8;
+    int ret = 0;
+    if ((ret = memcmp(a, b, bytes)))
+        return ret;
+    return (int)((a[bytes] & bitmask) - (b[bytes] & bitmask));
+}
+
+/* Compares 2 IPHC context entries */
+static int compare_ctx(void *a, void *b)
+{
+    struct iphc_ctx *ca = (struct iphc_ctx *)a;
+    struct iphc_ctx *cb = (struct iphc_ctx *)b;
+    return compare_prefix(ca->prefix.addr, cb->prefix.addr, ca->size);
+}
+
+PICO_TREE_DECLARE(CTXtree, compare_ctx);
+
+/* Searches in the context tree if there's a context entry available with the
+ * prefix of the IPv6 address */
+struct iphc_ctx * ctx_lookup(struct pico_ip6 addr)
+{
+    struct iphc_ctx test = { addr, 0, 0, 0, 0, NULL };
+    return pico_tree_findKey(&CTXtree, &test);
+}
+
+/* Looks up the context by ID, for decompression */
+struct iphc_ctx * ctx_lookup_id(uint8_t id)
+{
+    struct iphc_ctx *key = NULL;
+    struct pico_tree_node *i = NULL;
+
+    pico_tree_foreach(i, &CTXtree) {
+        key = i->keyValue;
+        if (key && id ==key->id)
+            return key;
+    }
```

Comment:
If it's only 16 elements, it shouldn't be too bad. Leave as-is. If it causes problems, we can still have a look at it then.